akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -540,27 +562,51 @@ private[stream] class ActorGraphInterpreter(_initial: GraphInterpreterShell) ext
     if (activeInterpreters.isEmpty) context.stop(self)
   }
 
+  private def shortCircuitBatch(): Unit = {
+    @tailrec
+    def shortCircuitBatchUtil(): Unit =
+      if (!shortCircuitBuffer.isEmpty) {
+        val input = shortCircuitBuffer.dequeue()
+        processEvent(input)
+        shortCircuitBatchUtil()
+      }
+    currentLimit = shortCircuitLimit
+    shortCircuitBatchUtil()
+  }
+
+  private def processEvent(b: BoundaryEvent): Unit = {
+    val shell = b.shell
+    if (!shell.isTerminated && (shell.isInitialized || tryInit(shell))) {
+      shell.receive(b)
+      if (shell.isTerminated) {
+        activeInterpreters -= shell
+        if (activeInterpreters.isEmpty && newShells.isEmpty) context.stop(self)
+      }
+    }
+  }
+
   override def receive: Receive = {
     case b: BoundaryEvent ⇒
-      val shell = b.shell
-      if (!shell.isTerminated && (shell.isInitialized || tryInit(shell))) {
-        shell.receive(b)
-        if (shell.isTerminated) {
-          activeInterpreters -= shell
-          if (activeInterpreters.isEmpty && newShells.isEmpty) context.stop(self)
-        }
-      }
-    case Resume ⇒ finishShellRegistration()
+      processEvent(b)
+      if (shortCircuitBuffer != null) shortCircuitBatch()
+
+    case Resume ⇒
+      finishShellRegistration()
+      if (shortCircuitBuffer != null) shortCircuitBatch()
+
     case StreamSupervisor.PrintDebugDump ⇒
-      println(s"activeShells:")
-      activeInterpreters.foreach { shell ⇒
-        println("  " + shell.toString.replace("\n", "\n  "))
-        shell.interpreter.dumpWaits()
-      }
-      println(s"newShells:")
-      newShells.foreach { shell ⇒
-        println("  " + shell.toString.replace("\n", "\n  "))
-        shell.interpreter.dumpWaits()
+      _initial.mat.synchronized {
```

Comment:
ah, okay; then a better solution is to construct the whole string an `println` just once